src/lib/util/heap.mli

```diff
@@ -28,48 +28,79 @@
 (*                                                                        *)
 (**************************************************************************)
 
-(** Integer heaps
+(** Heaps.
 
-    This modules define priority heaps over integers.
+    This modules define intrusive priority heaps (i.e. priority heaps where the
+    index of a given element in the map is stored on the element itself).
 *)
 
-(** {2 Integer heaps} *)
+(** {2 Ranked types}
 
-type t
-(** The type of heaps. *)
+    Elements that are put in a heap must implement the {!RankedType} interface.
+    This requires providing a (loose) comparison function, and allows to get and
+    set an `index` field on the element (to map back the element to its current
+    position in the heap).
 
-val init : int -> t
-(** Create a heap with the given initial size. *)
+    Note that this implies that each element can only be in a single heap at a
+    time. *)
 
-val in_heap : t -> int -> bool
-(** Heap membership function. *)
+module type RankedType = sig
+  type t
+  (** The type of the heap elements. *)
 
-val decrease : (int -> int -> bool) -> t -> int -> unit
-(** Decrease activity of the given integer.
-    TODO: document the comparison function ! *)
+  val index : t -> int
+  (** Index of the element in the heap. Returns -1 if the element is not in
+      the heap. *)
 
-val increase : (int -> int -> bool) -> t -> int -> unit
-(** Increase activity of the given integer.
-    TODO: document the comparison function ! *)
+  val set_index : t -> int -> unit
+  (** Update the element's index in the heap. *)
 
-val size : t -> int
-(** Returns the current size of the heap. *)
+  val compare : t -> t -> int
+  (** A total ordering function over the set elements.
+      This is a two-argument function [f] such that
+      [f e1 e2] is zero if the elements [e1] and [e2] are equal,
+      [f e1 e2] is strictly negative if [e1] is smaller than [e2],
+      and [f e1 e2] is strictly positive if [e1] is greater than [e2]. *)
+end
```

Comment:
As you taught me, doing arrays of pairs isn't perform in both memory and access as well as doing pairs of arrays. We should benchmark this code.

Reply:
I am not sure this matters much here because this is not about arrays of pairs vs pairs of arrays, it is about replacing an indirection through an array (`indices.(i)`) by a direct access (`a.hindex`) so there is only one level of indirection in each case, the data is just stored differently (and we also don't iterate linearly on the array). I am OK with running benchmarks tho.

Reply:
I think you should run benchmarks at least to check we don't break an implicit invariant. Maybe another part of the code used the fact that vectors were sparse. It's unlikely but still let's check it :)